### ISSUE TITLE
The current kubelet device plugin code could fail to remove some devices that are no longer advertised by a re-registered device plugin

### ISSUE THREAD

**Issue by jiayingz** [kubernetes/kubernetes#55179]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
Consider this corner case scenario: a device plugin registers with Kubelet for resourceA and lists two devices: dev1 and dev2. Then a new device plugin implementation is deployed that re-registers resourceA with Kubelet but somehow advertises with different device ids, say dev3 and dev4. With the current implementation, we will end up having dev1, dev2, dev3, and dev4 for resourceA incorrectly.

To fix this problem, one possible solution is to pass list of current devices in newEndpoint(). For the first registered device plugin, the list will be empty. For the subsequent re-registered plugin, the list will be what reported by its previous device plugin.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
The problem should be able to be reproduced with a unit test.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by jiayingz** [comment]
@lichuqiang do you want to work on this after your PR https://github.com/kubernetes/kubernetes/pull/52149?

**Comment by lichuqiang** [comment]
@jiayingz Yes, I'm glad to work on it

**Comment by lichuqiang** [comment]
I added a commit in #52149 to fix the issue. 
However, what I'm concerned about is by taking `devices` as a params of newEndpoint(), the func would depends on the bahavior of endpoints store in manager, thus, the scope of the lock here would extend, which may have impact on performance.
```
	// Associates the newly created endpoint with the corresponding resource name.
	// Stops existing endpoint if there is any.
	m.mutex.Lock()
	old, ok := m.endpoints[r.ResourceName]
	m.endpoints[r.ResourceName] = e
	glog.V(2).Infof("Registered endpoint %v", e)
	m.mutex.Unlock()
```
I thought to pass the devices after newEndpoint(), and before e.run(). 
```
	if ok && old != nil {
		// Pass devices of previous endpoint into re-registered one,
		// to avoid potential orphaned devices upon re-registration
		e.devices = old.devices
		old.stop()
	}
```
But this way, if the plugin re-register frequently enough, say a new endpoint has just been added in store, and before its devices assigned, another registation comes, then `old.devices` would be nil and the real devices might still be orphaned.

What's your opinion?

**Comment by ScorpioCPH** [comment]
@lichuqiang Hi, I update UT in this [PR](https://github.com/kubernetes/kubernetes/pull/55787) to try to reproduce this corner case, but it seems ok for now.

Maybe we need e2e test to reproduce this issue?

**Comment by jiayingz** [comment]
I think this issue is fixed by pr #52149. @lichuqiang can we close this issue?

**Comment by lichuqiang** [comment]
@jiayingz yes, please
